### PR TITLE
planner: forbid converting subqueries to inner-joins if the left and the right key have different collations

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
When trying to convert a sub-query to an inner-join, check the left key and the right key, and if they have different charsets or collations, forbid this converting.
When converting we will add a distinct-agg upon the right operator and making this distinct-agg consider collation seems tricky, so to keep it simple, we forbid this converting in this case.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -796,11 +796,22 @@ func (er *expressionRewriter) handleInSubquery(ctx context.Context, v *ast.Patte
 		er.err = err
 		return v, true
 	}
+
+	// If the leftKey and the rightKey have different charsets or collations, don't convert the sub-query to an inner-join
+	// since when converting we will add a distinct-agg upon the right child and this distinct-agg doesn't have the right collation.
+	// To keep it simple, we forbid this converting if they have different collations.
+	collFlag := true
+	lt, rt := lexpr.GetType(), rexpr.GetType()
+	if lt.EvalType() == types.ETString && rt.EvalType() == types.ETString &&
```
Comment (wjhuang2016) [inline planner/core/expression_rewriter.go]: Please consider `CompatibleCollate`, I think this function is enough, we don't need to check the charset any more.

Comment (qw4990) [reply planner/core/expression_rewriter.go]: fixed, PTAL!

**Comment by qw4990** [comment]
> Should it also be applied with `EXIST` subquery?

We don't convert exist subqueries to inner-joins when handling them(in the function `handleExistSubquery`), so it doesn't have this problem, but I still add some tests for it.

**Pull request merged**